pkg/storegateway/indexheader/index/symbols.go

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/tsdb/index/index.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors.
+
+package index
+
+import (
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"sort"
+	"unsafe"
+
+	"github.com/prometheus/prometheus/tsdb/index"
+
+	stream_encoding "github.com/grafana/mimir/pkg/storegateway/indexheader/encoding"
+)
+
+// The table gets initialized with sync.Once but may still cause a race
+// with any other use of the crc32 package anywhere. Thus we initialize it
+// before.
+var castagnoliTable *crc32.Table
+
+func init() {
+	castagnoliTable = crc32.MakeTable(crc32.Castagnoli)
+}
+
+type Symbols struct {
+	factory *stream_encoding.DecbufFactory
+
+	version     int
+	tableLength int
+	tableOffset int
+
+	offsets []int
+	seen    int
+}
+
+const symbolFactor = 32
+
+// NewSymbols returns a Symbols object for symbol lookups.
+func NewSymbols(factory *stream_encoding.DecbufFactory, version, offset int) (s *Symbols, err error) {
+	d := factory.NewDecbufAtChecked(offset, castagnoliTable)
+	defer factory.CloseWithErrCapture(&err, d, "read symbols")
+	if err := d.Err(); err != nil {
+		return nil, fmt.Errorf("decode symbol table: %w", d.Err())
+	}
+
+	s = &Symbols{
+		factory:     factory,
+		version:     version,
+		tableLength: d.Len() + 4, // NewDecbufAtChecked has already read the size of the table (4 bytes) by the time we get here.
+		tableOffset: offset,
+	}
+
+	origLen := d.Len()
+	cnt := d.Be32int()
+	basePos := 4
+	s.offsets = make([]int, 0, 1+cnt/symbolFactor)
+	for d.Err() == nil && s.seen < cnt {
+		if s.seen%symbolFactor == 0 {
+			s.offsets = append(s.offsets, basePos+origLen-d.Len())
+		}
+		d.UvarintBytes() // The symbol.
+		s.seen++
+	}
+
+	if d.Err() != nil {
+		return nil, d.Err()
+	}
+
+	return s, nil
+}
+
+func (s *Symbols) Lookup(o uint32) (sym string, err error) {
+	d := s.factory.NewDecbufAtUnchecked(s.tableOffset)
+	defer s.factory.CloseWithErrCapture(&err, d, "lookup symbol")
+	if err := d.Err(); err != nil {
+		return "", err
+	}
+
+	if s.version == index.FormatV2 {
+		if int(o) >= s.seen {
+			return "", fmt.Errorf("unknown symbol offset %d", o)
+		}
+		d.ResetAt(s.offsets[int(o/symbolFactor)])
+		// Walk until we find the one we want.
+		for i := o - (o / symbolFactor * symbolFactor); i > 0; i-- {
+			d.UvarintBytes()
```

Comment:
That's basically what `d.UvarintBytes()` does with the addition of `Peek()` that doesn't allocate. I'm going to leave it for now because this pattern is used in a few places including the existing `BinaryReader`.